File: android/app/src/main/java/de/artikelfinder/app/data/Suchtext.kt
package de.artikelfinder.app.data

import java.text.Normalizer
import java.util.Locale

/**
 * Normalisiert Suchbegriffe und Artikelnamen auf eine gemeinsame Form.
 *
 * Portiert aus `ArtikelFinder.Shared.Suchtext` — solange der Katalog auf dem Server
 * erzeugt und auf dem Gerät durchsucht wird, müssen beide Seiten identisch normalisieren.
 * Die Tests auf beiden Seiten prüfen dieselben Beispiele.
 */
object Suchtext {

    /** Schreibweise, die Deutsche beim Tippen ohne Umlaut-Taste erwarten. */
    private val umlaute = listOf(
        Triple('ä', "ae", "a"),
        Triple('ö', "oe", "o"),
        Triple('ü', "ue", "u"),
        Triple('ß', "ss", "ss"),
    )

    /** Kanonische Form eines Suchbegriffs: klein, ohne Satzzeichen, Umlaute ausgeschrieben. */
    fun normalisieren(eingabe: String?): String = aufbereiten(eingabe, kurzform = false)

    /**
     * Der Text, der in der Datenbank landet. Enthält zusätzlich die Kurzform der Umlaute,
     * damit sowohl "mueller" als auch "muller" den Artikel "Müller" findet.
     */
    fun fuerIndex(eingabe: String?): String {
        val lang = aufbereiten(eingabe, kurzform = false)
        val kurz = aufbereiten(eingabe, kurzform = true)
        return if (lang == kurz) lang else "$lang $kurz"
    }

    private fun aufbereiten(eingabe: String?, kurzform: Boolean): String {
        if (eingabe.isNullOrBlank()) return ""

        val ersetzt = StringBuilder(eingabe.length + 4)
        for (zeichen in eingabe.lowercase(Locale.GERMANY)) {
            val treffer = umlaute.firstOrNull { it.first == zeichen }
            if (treffer != null) {
                ersetzt.append(if (kurzform) treffer.third else treffer.second)
            } else {
                ersetzt.append(zeichen)
            }
        }

        // Verbleibende Akzente (é, ñ, …) auf den Basisbuchstaben zurückführen.
        val zerlegt = Normalizer.normalize(ersetzt, Normalizer.Form.NFD)
        val ergebnis = StringBuilder(zerlegt.length)
        var letzterWarTrenner = false

        for (zeichen in zerlegt) {
            when {
                Character.getType(zeichen) == Character.NON_SPACING_MARK.toInt() -> Unit

                zeichen.isLetterOrDigit() -> {
                    ergebnis.append(zeichen)
                    letzterWarTrenner = false
                }

                !letzterWarTrenner && ergebnis.isNotEmpty() -> {
                    ergebnis.append(' ')
                    letzterWarTrenner = true
                }
            }
        }

        return Normalizer.normalize(ergebnis.toString().trimEnd(), Normalizer.Form.NFC)
    }
}

/** Hilfsfunktionen für Barcodes, portiert aus `ArtikelFinder.Shared.Ean`. */
object Ean {

    /** Bringt eine EAN auf Normalform: nur Ziffern, führende Nullen bleiben erhalten. */
    fun normalisieren(eingabe: String?): String? {
        if (eingabe.isNullOrBlank()) return null
        val ziffern = eingabe.filter(Char::isDigit)
        return if (ziffern.length in 8..14) ziffern else null
    }

    /** Prüft die GTIN-Prüfziffer (EAN-8, UPC-A, EAN-13, GTIN-14). */
    fun pruefzifferKorrekt(ean: String?): Boolean {
        val normalisiert = normalisieren(ean) ?: return false
        if (normalisiert.length !in setOf(8, 12, 13, 14)) return false

        var summe = 0
        var gewicht = 3
        for (i in normalisiert.length - 2 downTo 0) {
            summe += (normalisiert[i] - '0') * gewicht
            gewicht = if (gewicht == 3) 1 else 3
        }

        return (10 - summe % 10) % 10 == normalisiert.last() - '0'
    }
}

File: backend/src/ArtikelFinder.Shared/Suchtext.cs
using System.Globalization;
using System.Text;

namespace ArtikelFinder.Shared;

/// <summary>
/// Normalisiert Suchbegriffe und Artikelnamen auf eine gemeinsame Form, damit die
/// Volltextsuche ohne datenbankspezifische Collations auskommt und auf SQLite wie auf
/// PostgreSQL identisch trifft.
/// </summary>
public static class Suchtext
{
    /// <summary>Schreibweise, die Deutsche beim Tippen ohne Umlaut-Taste erwarten.</summary>
    private static readonly (char Umlaut, string Lang, string Kurz)[] Umlaute =
    [
        ('ä', "ae", "a"),
        ('ö', "oe", "o"),
        ('ü', "ue", "u"),
        ('ß', "ss", "ss"),
    ];

    /// <summary>
    /// Kanonische Form eines Suchbegriffs: klein, ohne Satzzeichen, Umlaute ausgeschrieben.
    /// "Müller Käse!" wird zu "mueller kaese".
    /// </summary>
    public static string Normalisieren(string? eingabe) => Aufbereiten(eingabe, kurzform: false);

    /// <summary>
    /// Der Text, der in der Datenbank landet. Enthaelt zusaetzlich die Kurzform der
    /// Umlaute, damit sowohl "mueller" als auch "muller" den Artikel "Müller" finden —
    /// beide Schreibweisen sind im Alltag ueblich.
    /// </summary>
    public static string FuerIndex(string? eingabe)
    {
        var lang = Aufbereiten(eingabe, kurzform: false);
        var kurz = Aufbereiten(eingabe, kurzform: true);

        return string.Equals(lang, kurz, StringComparison.Ordinal) ? lang : $"{lang} {kurz}";
    }

    private static string Aufbereiten(string? eingabe, bool kurzform)
    {
        if (string.IsNullOrWhiteSpace(eingabe))
        {
            return string.Empty;
        }

        var ersetzt = new StringBuilder(eingabe.Length + 4);
        foreach (var zeichen in eingabe.ToLowerInvariant())
        {
            var gefunden = false;
            foreach (var (umlaut, lang, kurz) in Umlaute)
            {
                if (zeichen == umlaut)
                {
                    ersetzt.Append(kurzform ? kurz : lang);
                    gefunden = true;
                    break;
                }
            }

            if (!gefunden)
            {
                ersetzt.Append(zeichen);
            }
        }

        // Verbleibende Akzente (é, ñ, …) auf den Basisbuchstaben zurueckfuehren.
        var zerlegt = ersetzt.ToString().Normalize(NormalizationForm.FormD);
        var ergebnis = new StringBuilder(zerlegt.Length);
        var letzterWarTrenner = false;

        foreach (var zeichen in zerlegt)
        {
            if (CharUnicodeInfo.GetUnicodeCategory(zeichen) == UnicodeCategory.NonSpacingMark)
            {
                continue;
            }

            if (char.IsLetterOrDigit(zeichen))
            {
                ergebnis.Append(zeichen);
                letzterWarTrenner = false;
            }
            else if (!letzterWarTrenner && ergebnis.Length > 0)
            {
                // Satzzeichen und Mehrfach-Leerzeichen werden zu genau einem Trenner.
                ergebnis.Append(' ');
                letzterWarTrenner = true;
            }
        }

        return ergebnis.ToString().TrimEnd().Normalize(NormalizationForm.FormC);
    }
}
